File: Samples/Verbum/how-to-embed-a-barcode-into-a-document-e20025-11.1.6-/CS/Extensions/BarCode/Helpers/Code128GeneratorHelper.cs
namespace BizPad {
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using DevExpress.XtraRichEdit.Fields;
    using DevExpress.XtraRichEdit.Model;
    using System.IO;
    using DevExpress.XtraCharts;
    using System.Drawing.Imaging;
    using DevExpress.XtraRichEdit.Utils;
    using System.Text;
    using System.Drawing;
    using DevExpress.XtraPrinting.BarCode;
    using DevExpress.XtraPrinting.BarCode.Native;
    using System.Collections;
    using DevExpress.XtraPrinting;
    using System.Threading;
    using DevExpress.XtraPrinting.Localization;
    using DevExpress.Utils.Text;

    internal class Code128GeneratorHelper : Code128Generator, IBarCodeGenerator {
        static Dictionary<string, Code128Charset> Subtypes = new Dictionary<string, Code128Charset>();
        static Code128GeneratorHelper() {
            Subtypes.Add("128a", Code128Charset.CharsetA);
            Subtypes.Add("128b", Code128Charset.CharsetB);
            Subtypes.Add("128c", Code128Charset.CharsetC);
        }
        
        public Code128GeneratorHelper() {
        }

        protected Code128GeneratorHelper(Code128Generator source)
            : base(source) {
        }

        void IBarCodeGenerator.Initialize(InstructionCollection switches) {
            string p = switches.GetString("p") ?? String.Empty;

            Code128Charset subtype = Code128Charset.CharsetB;
            if (Subtypes.ContainsKey(p.ToLower())) {
                subtype = Subtypes[p.ToLower()];
            }

            this.CharacterSet = subtype;
        }

        public bool SupressAutoTextAlignment() {
            return false;
        }

        string IBarCodeGenerator.GetValidCharSet() {
            return base.GetValidCharSet();
        }

        ArrayList IBarCodeGenerator.MakeBarCodePattern(string text) {
            return base.MakeBarCodePattern(text);
        }

        void IBarCodeGenerator.DrawBarCode(IGraphics gr, RectangleF barBounds, RectangleF textBounds, IBarCodeData data, float xModule, float yModule) {
            base.DrawBarCode(gr, barBounds, textBounds, data, xModule, yModule);
        }

        string IBarCodeGenerator.MakeDisplayText(string text) {
            return base.MakeDisplayText(text);
        }

        string IBarCodeGenerator.FormatText(string text) {
            return base.FormatText(text);
        }

        bool IBarCodeGenerator.IsValidPattern(ArrayList pattern) {
            return base.IsValidPattern(pattern);
        }

        bool IBarCodeGenerator.IsValidText(string text) {
            return BarCodeGeneratorHelper.IsValidText(this, text);
        }

        bool IBarCodeGenerator.IsValidTextFormat(string text) {
            return base.IsValidTextFormat(text);
        }

        float IBarCodeGenerator.CalcBarCodeWidth(ArrayList pattern) {
            return base.CalcBarCodeWidth(pattern);
        }

        float IBarCodeGenerator.CalcBarCodeHeight(ArrayList pattern) {
            return base.CalcBarCodeHeight(pattern);
        }
    }
}

File: Samples/Verbum/how-to-embed-a-barcode-into-a-document-e20025-11.1.6-/VB/Extensions/BarCode/Helpers/Code128GeneratorHelper.vb
Imports Microsoft.VisualBasic
	Imports System
	Imports System.Collections.Generic
	Imports System.Linq
	Imports DevExpress.XtraRichEdit.Fields
	Imports DevExpress.XtraRichEdit.Model
	Imports System.IO
	Imports DevExpress.XtraCharts
	Imports System.Drawing.Imaging
	Imports DevExpress.XtraRichEdit.Utils
	Imports System.Text
	Imports System.Drawing
	Imports DevExpress.XtraPrinting.BarCode
	Imports DevExpress.XtraPrinting.BarCode.Native
	Imports System.Collections
	Imports DevExpress.XtraPrinting
	Imports System.Threading
	Imports DevExpress.XtraPrinting.Localization
	Imports DevExpress.Utils.Text
Namespace BizPad

	Friend Class Code128GeneratorHelper
		Inherits Code128Generator
		Implements IBarCodeGenerator
		Private Shared Subtypes As New Dictionary(Of String, Code128Charset)()
		Shared Sub New()
			Subtypes.Add("128a", Code128Charset.CharsetA)
			Subtypes.Add("128b", Code128Charset.CharsetB)
			Subtypes.Add("128c", Code128Charset.CharsetC)
		End Sub

		Public Sub New()
		End Sub

		Protected Sub New(ByVal source As Code128Generator)
			MyBase.New(source)
		End Sub

		Private Sub Initialize(ByVal switches As InstructionCollection) Implements IBarCodeGenerator.Initialize
			Dim p As String = If(switches.GetString("p"), String.Empty)

			Dim subtype As Code128Charset = Code128Charset.CharsetB
			If Subtypes.ContainsKey(p.ToLower()) Then
				subtype = Subtypes(p.ToLower())
			End If

			Me.CharacterSet = subtype
		End Sub

		Public Function SupressAutoTextAlignment() As Boolean Implements IBarCodeGenerator.SupressAutoTextAlignment
			Return False
		End Function

		Private Function GetValidCharSet() As String Implements IBarCodeGenerator.GetValidCharSet
			Return MyBase.GetValidCharSet()
		End Function

		Private Function MakeBarCodePattern(ByVal text As String) As ArrayList Implements IBarCodeGenerator.MakeBarCodePattern
			Return MyBase.MakeBarCodePattern(text)
		End Function

		Private Sub DrawBarCode(ByVal gr As IGraphics, ByVal barBounds As RectangleF, ByVal textBounds As RectangleF, ByVal data As IBarCodeData, ByVal xModule As Single, ByVal yModule As Single) Implements IBarCodeGenerator.DrawBarCode
			MyBase.DrawBarCode(gr, barBounds, textBounds, data, xModule, yModule)
		End Sub

		Private Function MakeDisplayText(ByVal text As String) As String Implements IBarCodeGenerator.MakeDisplayText
			Return MyBase.MakeDisplayText(text)
		End Function

		Private Function FormatText(ByVal text As String) As String Implements IBarCodeGenerator.FormatText
			Return MyBase.FormatText(text)
		End Function

		Private Function IsValidPattern(ByVal pattern As ArrayList) As Boolean Implements IBarCodeGenerator.IsValidPattern
			Return MyBase.IsValidPattern(pattern)
		End Function

		Private Function IsValidText(ByVal text As String) As Boolean Implements IBarCodeGenerator.IsValidText
			Return BarCodeGeneratorHelper.IsValidText(Me, text)
		End Function

		Private Function IsValidTextFormat(ByVal text As String) As Boolean Implements IBarCodeGenerator.IsValidTextFormat
			Return MyBase.IsValidTextFormat(text)
		End Function

		Private Function CalcBarCodeWidth(ByVal pattern As ArrayList) As Single Implements IBarCodeGenerator.CalcBarCodeWidth
			Return MyBase.CalcBarCodeWidth(pattern)
		End Function

		Private Function CalcBarCodeHeight(ByVal pattern As ArrayList) As Single Implements IBarCodeGenerator.CalcBarCodeHeight
			Return MyBase.CalcBarCodeHeight(pattern)
		End Function
	End Class
End Namespace
